src/librustdoc/html/static/js/search.js

```diff
@@ -167,33 +503,142 @@ window.initSearch = function(rawSearchIndex) {
     }
 
     /**
-     * Executes the query and returns a list of results for each results tab.
-     * @param  {Object}        query          - The user query
-     * @param  {Array<string>} searchWords    - The list of search words to query against
-     * @param  {string}        [filterCrates] - Crate to search in
-     * @return {{
-     *   in_args: Array<?>,
-     *   returned: Array<?>,
-     *   others: Array<?>
-     * }}
+     * Parses the query.
+     *
+     * The supported syntax by this parser is as follow:
+     *
+     * ident = *(ALPHA / DIGIT / "_")
+     * path = ident *(DOUBLE-COLON ident)
+     * arg = path [generics]
+     * arg-without-generic = path
+     * type-sep = COMMA/WS *(COMMA/WS)
+     * nonempty-arg-list = arg *(type-sep arg) *(COMMA/WS)
+     * nonempty-arg-list-without-generics = arg-without-generic *(type-sep arg-without-generic)
+     *                                      *(COMMA/WS)
+     * generics = OPEN-ANGLE-BRACKET *WS [ nonempty-arg-list-without-generics ] *WS
+     *            CLOSE-ANGLE-BRACKET
+     * return-args = RETURN-ARROW *WS nonempty-arg-list
```

Comment:
Found another couple spots where the reference impl and eBNF don't match the JS impl:

    a->a

This eBNF has no construct for a query with parameters and a return value, yet the parser implements it.

    a->

This eBNF says that empty return args are not allowed, but the implementation parses them without error.

Reply:
> a->a

The parser is correct here, and the eBNF is wrong: this is part of the documented query syntax.

> a->

On nightly, searching with this syntax unconditionally returns the same set of results, regardless of `a`'s value: https://doc.rust-lang.org/nightly/std/string/struct.String.html?search=asfksldjfsfasd-%3E. So searching for any concrete `a`, e.g. `u8 ->` doesn't work. We can consider this syntax "not currently valid." I'd like to keep it that way, and reject it in the new and improved parser.

Reply:
I thought `u8 ->` was supposed to search for `fn(..., x: u8, ...) -> {anything}`?

Reply:
@camelid For now it doesn't. Might be a later improvement though!